packages/nextjs/src/utils/handlers.ts

```diff
@@ -24,73 +30,114 @@ export const withSentry = (handler: NextApiHandler): WrappedNextApiHandler => {
-              traceparentData = extractTraceparentData(req.headers['sentry-trace'] as string);
-              logger.log(`[Tracing] Continuing trace ${traceparentData?.traceId}.`);
-            }
+      if (currentScope) {
+        currentScope.addEventProcessor(event => parseRequest(event, req));
+
+        if (hasTracingEnabled()) {
+          // If there is a trace header set, extract the data from it (parentSpanId, traceId, and sampling decision)
+          let traceparentData;
+          if (req.headers && isString(req.headers['sentry-trace'])) {
+            traceparentData = extractTraceparentData(req.headers['sentry-trace'] as string);
+            logger.log(`[Tracing] Continuing trace ${traceparentData?.traceId}.`);
+          }
 
-            const url = `${req.url}`;
-            // pull off query string, if any
-            let reqPath = stripUrlQueryAndFragment(url);
-            // Replace with placeholder
-            if (req.query) {
-              // TODO get this from next if possible, to avoid accidentally replacing non-dynamic parts of the path if
-              // they match dynamic parts
-              for (const [key, value] of Object.entries(req.query)) {
-                reqPath = reqPath.replace(`${value}`, `[${key}]`);
-              }
+          const url = `${req.url}`;
+          // pull off query string, if any
+          let reqPath = stripUrlQueryAndFragment(url);
+          // Replace with placeholder
+          if (req.query) {
+            // TODO get this from next if possible, to avoid accidentally replacing non-dynamic parts of the path if
+            // they match dynamic parts
+            for (const [key, value] of Object.entries(req.query)) {
+              reqPath = reqPath.replace(`${value}`, `[${key}]`);
             }
-            const reqMethod = `${(req.method || 'GET').toUpperCase()} `;
-
-            const transaction = startTransaction(
-              {
-                name: `${reqMethod}${reqPath}`,
-                op: 'http.server',
-                ...traceparentData,
-              },
-              // extra context passed to the `tracesSampler`
-              { request: req },
-            );
-            currentScope.setSpan(transaction);
           }
+          const reqMethod = `${(req.method || 'GET').toUpperCase()} `;
+
+          const transaction = startTransaction(
+            {
+              name: `${reqMethod}${reqPath}`,
+              op: 'http.server',
+              ...traceparentData,
+            },
+            // extra context passed to the `tracesSampler`
+            { request: req },
+          );
+          currentScope.setSpan(transaction);
+
+          // save a link to the scope on the response, so that even if there's an error (landing us outside of
+          // the domain), we can still finish the transaction and attach the correct data to it
+          (res as ScopedResponse).__sentryScope = currentScope;
```

Comment:
Should we just clone the hub instead? Matches the hub per request model.

Reply:
Could do. We don't currently have a `clone` method on `Hub`, and `Hub.run`/`Hub.makeMain` are going to suffer from the same problem that `withScope` and `pushScope` suffer from, which is that none of them is designed in an async-proof way (which is to say, in none of the cases can we guarantee that the current hub/top scope won't have been changed in a different thread* before we go to restore the previous state of things). One other potential pitfall I see (haven't yet figured out if it would be a problem in practice) is that normally each hub has its own copy of every integration, whereas if we follow the current cloning method for scopes, it'd end up just being multiple hubs pointing to a single integration instance. As I say, might or might not be a problem, but something we'd have to think through.

Is there an advantage you see in this case to using a hub rather than a scope?

\*I know that JS is technically single-threaded, but hopefully you get what I mean

Reply:
Pinging @rhcarvalho so he can add his thoughts.

Reply:
From what I glanced, we only need a reference to the `transaction` / `span`, we don't really need anything else from the `Scope`.

If anything, I'd store only the `span` knowing that other contextual data might be incorrect (fixing that for the general case requires more involved work, something that we put aside for now).